### PR TITLE
Bugfix adjust port after binding.

### PR DESCRIPTION
**Pull request by boaks**
Adjust port after binding (TCP server).

Signed-off-by: Achim Kraus <achim.kraus@bosch-si.com>

### CODE REVIEW

**Comment by boaks** [comment]
In my system netty sets the bind address to IPv6 in "any" is used, which causes some trouble. Therefore I just pick the port, if not already provided .

File: element-connector/src/main/java/org/eclipse/californium/elements/tcp/TcpServerConnector.java [hunk 1]
```diff
@@ -82,17 +97,33 @@ public TcpServerConnector(InetSocketAddress localAddress, int numberOfThreads, i
...
-
-			@Override
-			public void operationComplete(ChannelFuture future) throws Exception {
-				InetSocketAddress listenAddress = (InetSocketAddress) future.channel().localAddress();
-				listenUri = getListenUri(listenAddress);
+		ChannelFuture channelFuture = bootstrap.bind(localAddress).syncUninterruptibly();
+
+		if (0 == localAddress.getPort()) {
+			// get assigned port
+			channelFuture.addListener(new ChannelFutureListener() {
```
Comment (joemag1) [inline element-connector/src/main/java/org/eclipse/californium/elements/tcp/TcpServerConnector.java]: Any value in listening asynchronously, since you already call syncUninterruptibly above?

Comment (boaks) [inline element-connector/src/main/java/org/eclipse/californium/elements/tcp/TcpServerConnector.java]: Sure, the value is "no need to read the javadoc" :-).

I wass just miss lead by the introduction of adding the listener before.
Thanks for the hint, it makes the implementation much simpler!

File: element-connector/src/main/java/org/eclipse/californium/elements/tcp/TcpServerConnector.java [hunk 2]
```diff
@@ -82,17 +85,17 @@ public TcpServerConnector(InetSocketAddress localAddress, int numberOfThreads, i
...
-			public void operationComplete(ChannelFuture future) throws Exception {
-				InetSocketAddress listenAddress = (InetSocketAddress) future.channel().localAddress();
-				listenUri = getListenUri(listenAddress);
-			}
-		});
+		ChannelFuture channelFuture = bootstrap.bind(localAddress).syncUninterruptibly();
+
+		if (channelFuture.isSuccess() && 0 == localAddress.getPort()) {
+			InetSocketAddress listenAddress = (InetSocketAddress) channelFuture.channel().localAddress();
+			localAddress = new InetSocketAddress(localAddress.getAddress(), listenAddress.getPort());
```
Comment (sophokles73) [inline element-connector/src/main/java/org/eclipse/californium/elements/tcp/TcpServerConnector.java]: wouldn't the following be sufficient:
```
localAddress = (InetSocketAddress) channelFuture.channel().localAddress();
listenUri = getListenUri(localAddress);
```

File: element-connector/src/main/java/org/eclipse/californium/elements/tcp/TcpServerConnector.java [hunk 3]
```diff
@@ -82,17 +85,17 @@ public TcpServerConnector(InetSocketAddress localAddress, int numberOfThreads, i
...
 		// Start the server.
-		bootstrap.bind(localAddress).syncUninterruptibly().addListener(new ChannelFutureListener() {
-
-			@Override
-			public void operationComplete(ChannelFuture future) throws Exception {
-				InetSocketAddress listenAddress = (InetSocketAddress) future.channel().localAddress();
-				listenUri = getListenUri(listenAddress);
-			}
-		});
+		ChannelFuture channelFuture = bootstrap.bind(localAddress).syncUninterruptibly();
```
Comment (sophokles73) [inline element-connector/src/main/java/org/eclipse/californium/elements/tcp/TcpServerConnector.java]: I assume that `syncUninterruptibly()` means something like: "do it now and wait for the result", right?

File: element-connector/src/main/java/org/eclipse/californium/elements/tcp/TcpServerConnector.java (hunk 2 above)
Comment (boaks) [inline element-connector/src/main/java/org/eclipse/californium/elements/tcp/TcpServerConnector.java]: Using that results in:

     coap+tcp://0:0:0:0:0:0:0:0:5683

and using that as (leshan example):

        this.registrationEndpointUri = registrationEndpointUri;
        this.registrationEndpointAddress = new InetSocketAddress(registrationEndpointUri.getHost(), registrationEndpointUri.getPort());

results in:

    java.lang.IllegalArgumentException: hostname can't be null
	at java.net.InetSocketAddress.checkHost(InetSocketAddress.java:149) ~[?:1.8.0_51]
	at java.net.InetSocketAddress.<init>(InetSocketAddress.java:216) ~[?:1.8.0_51]

So it seems, IPv6 has some open issues. Therefore I just replaced the port.
I had a comment on that, but on the reworks that comment seems to be deleted.
Alternatively, IPv6 usage may be consequently supported. But this seems to be much out of this scope.

Comment (boaks) [inline element-connector/src/main/java/org/eclipse/californium/elements/tcp/TcpServerConnector.java]: Analysing UDPConnector, I'm wondering, what the "exact" definition of "restart" with port 0 should be. Should it start on the previous port or could it alos chose a new one? UDPConnector preserves the 0 and would therefore potentially reopen a new port. Is that intended?

Comment (sophokles73) [inline element-connector/src/main/java/org/eclipse/californium/elements/tcp/TcpServerConnector.java]: Oh, now I see that you are using `localAddress.getAddress()` instead of `listenAddress.getAddress()`.

The problem lies in the `getListenUri()` method. We need to distinguish between `Inet4Address` and `Inet6Address` because the URI constructor requires IPv6 literal addresses to be enclosed in square brackets:

```
private URI getListenUri(final InetSocketAddress listenAddress) {

  InetAddress address = listenAddress.getAddress();
  String addressString = address.getHostString();
  if (address instanceof Inet6Address) {
    addressString = String.format("[%s]", addressString);
  }
      return URI.create(String.format("%s://%s:%d", getSupportedScheme(), addressString, address.getPort()));
}
```

Comment (boaks) [inline element-connector/src/main/java/org/eclipse/californium/elements/tcp/TcpServerConnector.java]: Hm, in this PR I would like to got for "the conservative solution" and just replace the port.
I'm afraid, that changing towards an IPv6 address would introduce more problems outsite.
Additionally we may add this InetSocketAddress to URI in a common util class.

Comment (sophokles73) [inline element-connector/src/main/java/org/eclipse/californium/elements/tcp/TcpServerConnector.java]: Sounds good. Could you please open an issue for it so that we do not forget it?

Comment (sophokles73) [inline element-connector/src/main/java/org/eclipse/californium/elements/tcp/TcpServerConnector.java]: > Analysing UDPConnector, I'm wondering, what the "exact" definition of "restart" with port 0 should be. Should it start on the previous port or could it alos chose a new one? UDPConnector preserves the 0 and would therefore potentially reopen a new port. Is that intended?

Is this related to this PR? If not and you suspect a problem with the current behavior, please open another issue.

**Comment by boaks** [comment]
Created two issues for the left stuff (IPv6 and wanted behavior on restart with port 0).
I added also a comment to just replace the port.
So I hope the requested changes are done.

With that commit the PR #148 could also be simplified a bit. So if merged, I will adjust the other PR also.
